typescript/cli/src/context/altvm.ts

```diff
@@ -0,0 +1,196 @@
+    }
+
+    const { gas } = ALT_VM_SUPPORTED_PROTOCOLS[protocol]!;
+    return gas;
+  }
+}
+
+export class AltVMProviderFactory
+  extends AltVMFactory
+  implements AltVM.IProviderFactory
+{
+  private readonly metadataManager: ChainMetadataManager;
+
+  constructor(metadataManager: ChainMetadataManager) {
+    super();
+
+    this.metadataManager = metadataManager;
+  }
+
+  public async get(chain: string): Promise<AltVM.IProvider> {
+    const metadata = this.metadataManager.getChainMetadata(chain);
+
+    if (!this.supports(metadata.protocol)) {
+      throw new Error(
+        `Chain ${chain} with protocol type ${metadata.protocol} not supported in AltVM`,
+      );
+    }
+
+    const { provider } = ALT_VM_SUPPORTED_PROTOCOLS[metadata.protocol]!;
+    return provider.connect(metadata.rpcUrls.map((rpc) => rpc.http));
+  }
+}
+
+export class AltVMSignerFactory
+  extends AltVMFactory
+  implements AltVM.ISignerFactory
+{
+  private readonly metadataManager: ChainMetadataManager;
+  private readonly chains: ChainMap<AltVM.ISigner>;
+
+  private constructor(
+    metadataManager: ChainMetadataManager,
+    chains: ChainMap<AltVM.ISigner>,
+  ) {
+    super();
+
+    this.metadataManager = metadataManager;
+    this.chains = chains;
+  }
+
+  public get(chain: string): AltVM.ISigner {
+    const protocol = this.metadataManager.getProtocol(chain);
+
+    if (!this.supports(protocol)) {
+      throw new Error(
+        `Chain ${chain} with protocol type ${protocol} not supported in AltVM`,
+      );
+    }
+
+    if (!this.chains[chain]) {
+      throw new Error(`AltVM was not initialized with chain ${chain}`);
+    }
+
+    return this.chains[chain];
+  }
+
+  private static async loadPrivateKey(
+    keyByProtocol: SignerKeyProtocolMap,
+    strategyConfig: Partial<ExtendedChainSubmissionStrategy>,
+    protocol: ProtocolType,
+    chain: string,
+  ): Promise<string> {
+    // 1. First try to get private key from --key.{protocol} flag
+    if (keyByProtocol[protocol]) {
+      return keyByProtocol[protocol]!;
+    }
+
+    // 2. If no key flag was provided we check if a strategy config
+    // was provided for our chain where we can read our private key
+    if (strategyConfig[chain]) {
+      const rawConfig = strategyConfig[chain]!.submitter;
+      if (!isJsonRpcSubmitterConfig(rawConfig)) {
+        throw new Error(
+          `found unknown submitter in strategy config for chain ${chain}`,
+        );
+      }
+
+      if (!rawConfig.privateKey) {
+        throw new Error(
+          `missing private key in strategy config for chain ${chain}`,
+        );
+      }
+
+      return rawConfig.privateKey;
+    }
+
+    // 3. Finally, if no key flag or strategy was provided we prompt the user
+    // for the private key
+    return password({
+      message: `Please enter the private key for chain ${chain} (will be re-used for other chains with the same protocol type)`,
+    });
+  }
```

Comment:
I think that one thing that we are loosing here compared to the existing logic is the ability reuse private keys when they are provided for a chain and that same key could be reused for another one as they share the same protocol type which would cause once again the user to be prompted multiple times for the same pk as you are not caching it by protocol here even if the message says otherwise